### PR TITLE
Add Orleans in backend and add some new APIs

### PR DESCRIPTION
**Pull request by jamesmontemagno**

### CODE REVIEW

**Comment by bradygaster** [comment]
Lgtm! Mobile, scrolled through, but the mobile app doesn't allow me to approve afaict. All good

File: src/Services/ListenTogether/ListenTogether.Application/Grains/RoomGrain.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
+﻿using ListenTogether.Application.Interfaces;
+using ListenTogether.Domain;
+using Microsoft.Extensions.Logging;
+using Orleans;
+using Orleans.Runtime;
+
+namespace ListenTogether.Application.Grains
+{
+    public class RoomGrain : Grain, IRoomGrain
```
Comment (IEvangelist) [inline src/Services/ListenTogether/ListenTogether.Application/Grains/RoomGrain.cs:9]:
```suggestion
    public sealed class RoomGrain : Grain, IRoomGrain
```

File: src/Services/ListenTogether/ListenTogether.Application/Grains/RoomGrain.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        IPersistentState<Room> Room { get; set; }
+        public ILogger<RoomGrain> Logger { get; }
```
Comment (IEvangelist) [inline src/Services/ListenTogether/ListenTogether.Application/Grains/RoomGrain.cs:12]:
```suggestion
        private readonly IPersistentState<Room> _room;
        private readonly ILogger<RoomGrain> _logger;
```

These are better being `readonly` fields, unless there is a good reason to expose them.

File: src/Services/ListenTogether/ListenTogether.Application/Grains/RoomGrain.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+        public RoomGrain([PersistentState("room", "roomStorage")] IPersistentState<Room> room,
+            ILogger<RoomGrain> logger)
+        {
+            Room = room;
+            Logger = logger;
```
Comment (IEvangelist) [inline src/Services/ListenTogether/ListenTogether.Application/Grains/RoomGrain.cs:18]:
```suggestion
            _room = room;
            _logger = logger;
```

File: src/Services/ListenTogether/ListenTogether.Application/Grains/RoomGrain.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        public async Task<Room> JoinRoom(string connectionId, string userName)
+        {
+            Logger.LogInformation($"User {userName} is joining {Room.State.Code}, listening to {Room.State.Episode.Show.Title}");
+            Room.State.AddUser(new User(connectionId, userName));
+            await Room.WriteStateAsync();
+            return Room.State;
```
Comment (IEvangelist) [inline src/Services/ListenTogether/ListenTogether.Application/Grains/RoomGrain.cs:26]:
```suggestion
            _logger.LogInformation($"User {userName} is joining {_room.State.Code}, listening to {_room.State.Episode.Show.Title}");
            _room.State.AddUser(new User(connectionId, userName));
            await _room.WriteStateAsync();
            return _room.State;
```

File: src/Services/ListenTogether/ListenTogether.Application/Grains/RoomGrain.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            Logger.LogInformation($"User {user.Name} is joining {Room.State.Code}, listening to {Room.State.Episode.Show.Title}");
+            Room.State.RemoveUser(connectionId);
+            await Room.WriteStateAsync();
+
+            if(Room.State.Users.Count == 0)
+            {
+                base.DeactivateOnIdle();
+            }
+
+            return Room.State;
```
Comment (IEvangelist) [inline src/Services/ListenTogether/ListenTogether.Application/Grains/RoomGrain.cs:41]:
```suggestion
            var user = _room.State.Users.First(_ => _.ConnectionId == connectionId);
            _logger.LogInformation($"User {user.Name} is joining {_room.State.Code}, listening to {_room.State.Episode.Show.Title}");
            _room.State.RemoveUser(connectionId);
            await _room.WriteStateAsync();

            if (_room.State.Users.Count is 0)
            {
                base.DeactivateOnIdle();
            }

            return _room.State;
```

File: src/Services/ListenTogether/ListenTogether.Application/Grains/RoomGrain.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        public async Task SetRoom(Room room)
+        {
+            Room.State = room;
+            await Room.WriteStateAsync();
```
Comment (IEvangelist) [inline src/Services/ListenTogether/ListenTogether.Application/Grains/RoomGrain.cs:47]:
```suggestion
            _room.State = room;
            await _room.WriteStateAsync();
```

File: src/Services/ListenTogether/ListenTogether.Application/Grains/RoomGrain.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        public Task<Room> UpdatePlayerState(TimeSpan seconds, PlayerState playerState)
+        {
+            Logger.LogInformation($"Updating player state to {playerState} at {seconds}.");
+            Room.State.UpdatePlayerState(seconds, playerState);
+            return Task.FromResult(Room.State);
```
Comment (IEvangelist) [inline src/Services/ListenTogether/ListenTogether.Application/Grains/RoomGrain.cs:54]:
```suggestion
            _logger.LogInformation($"Updating player state to {playerState} at {seconds}.");
            _room.State.UpdatePlayerState(seconds, playerState);
            return Task.FromResult(_room.State);
```

File: src/Services/ListenTogether/ListenTogether.Application/Interfaces/IRoomGrain.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+namespace ListenTogether.Application.Interfaces
+{
+    public interface IRoomGrain : IGrainWithStringKey
+    {
+        Task SetRoom(Room room);
+        Task<Room> JoinRoom(string connectionId, string userName);
+        Task<Room> LeaveRoom(string connectionId);
+        Task<Room> UpdatePlayerState(TimeSpan seconds, PlayerState playerState);
+    }
+}
```
Comment (IEvangelist) [inline src/Services/ListenTogether/ListenTogether.Application/Interfaces/IRoomGrain.cs:13]:
```suggestion
namespace ListenTogether.Application.Interfaces;

public interface IRoomGrain : IGrainWithStringKey
{
    Task SetRoom(Room room);
    Task<Room> JoinRoom(string connectionId, string userName);
    Task<Room> LeaveRoom(string connectionId);
    Task<Room> UpdatePlayerState(TimeSpan seconds, PlayerState playerState);
}
```

I vote for using file-scoped namespaces, what do you think?

File: src/Services/ListenTogether/ListenTogether.Domain/Room.cs [hunk 9]
```diff
@@ -6,7 +6,10 @@ public class Room
     private const string ValidCharacters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
     private readonly Random _randGenerator = new();
```
Comment (IEvangelist) [inline src/Services/ListenTogether/ListenTogether.Domain/Room.cs:7]:
```suggestion
    private readonly Random _randGenerator = Random.Shared;
```

File: src/Services/Podcasts/Podcast.Infrastructure/Data/Models/UserSubmittedFeed.cs [hunk 10]
```diff
@@ -0,0 +1,17 @@
+﻿namespace Podcast.Infrastructure.Data.Models;
+
+public class UserSubmittedFeed
```
Comment (IEvangelist) [inline src/Services/Podcasts/Podcast.Infrastructure/Data/Models/UserSubmittedFeed.cs:3]: Could this be a `record`, does EF support that?

File: src/Services/Podcasts/Podcast.Ingestion.Worker/PodcastIngestionHandler.cs [hunk 11]
```diff
@@ -27,34 +27,36 @@ public PodcastIngestionHandler(PodcastDbContext podcastDbContext, IFeedClient fe
...
     {
-        var isAcceptedTopic = AcceptedTopics.Any(topic => title.Contains(topic, StringComparison.InvariantCultureIgnoreCase));
-        if (!isAcceptedTopic) return;
-
-        var isExistingShow = await _podcastDbContext.Shows.AnyAsync(show => show.Title == title, stoppingToken);
-        if (isExistingShow) return;
-
-        var feed = new Feed(Guid.NewGuid(), url);
+        _logger.LogInformation($"The show {title} at {url} was received by the ingestion worker.");
+        var isExistingShow = await _podcastDbContext.Feeds.AnyAsync(feed => feed.Url.ToLower() == url.ToLower(), stoppingToken);
```
Comment (IEvangelist) [inline src/Services/Podcasts/Podcast.Ingestion.Worker/PodcastIngestionHandler.cs:31]:
```suggestion
        var isExistingShow = await _podcastDbContext.Feeds.AnyAsync(feed => string.Equals(feed.Url, url, StringComparison.OrdinalIgnoreCase), stoppingToken);
```

File: src/Services/Podcasts/Podcast.Ingestion.Worker/PodcastIngestionHandler.cs [hunk 12]
```diff
@@ -27,34 +27,36 @@ public PodcastIngestionHandler(PodcastDbContext podcastDbContext, IFeedClient fe
...
+                categoryStringArray += $"{cat},";
+            }
+
+            if(categoryStringArray.Contains(","))
+            {
+                categoryStringArray = categoryStringArray.Substring(0, categoryStringArray.LastIndexOf(','));
+            }
+
+            // Must be manually approved
+            var userFeed = new UserSubmittedFeed(title, url, categoryStringArray);
```
Comment (IEvangelist) [inline src/Services/Podcasts/Podcast.Ingestion.Worker/PodcastIngestionHandler.cs:52]:
```suggestion
            // Must be manually approved
            var userFeed = new UserSubmittedFeed(
                title, url, string.Join(",", feedCategories));
```

**Review by IEvangelist — COMMENTED** [review]
I have some suggestion to improve this.

File: src/Services/Podcasts/Podcast.Infrastructure/Data/Models/UserSubmittedFeed.cs (hunk 10 above)
Comment (jamesmontemagno) [reply src/Services/Podcasts/Podcast.Infrastructure/Data/Models/UserSubmittedFeed.cs:3]: i think not recommended at this time from my understanding.

**Comment by jamesmontemagno** [comment]
@bradygaster  @JeremyLikness @IEvangelist  anything else or is this good to go?

File: docker-compose.override.yml [hunk 13]
```diff
@@ -35,6 +35,7 @@ services:
       - ASPNETCORE_URLS=http://+:80
       - NetPodcastApi__BaseAddress=http://podcast.api
       - ConnectionStrings__ListenTogetherDb=Server=podcast.db;Database=ListenTogether;User Id=sa;Password=Pass@word
```
Comment (IEvangelist) [inline docker-compose.override.yml:37]: I don't think we should have passwords, or connection strings in source control.

**Review by IEvangelist — APPROVED** [review]
@jamesmontemagno Looks good to me, other than the sensitive info in the config/settings.

File: docker-compose.override.yml (hunk 13 above)
Comment (bradygaster) [reply docker-compose.override.yml:37]: if it's only for the local scenario i hate it less than if it's "guidance," but i do have a modicum of concern that we'd send the message that hard-coding strings is a good idea, and we know it isn't. Could we compromise and maybe add a .env file with guidance on how to fill it in?

**Review by bradygaster — APPROVED** [review]
As mentioned in the comment, I think the best compromise would be the .env part. The Orleans stuff I'm good with. :)

File: docker-compose.override.yml (hunk 13 above)
Comment (jamesmontemagno) [reply docker-compose.override.yml:37]: yeah I like that approach. someone want to do that ;)

**Pull request merged**